DOC FM => TeamMailboxe & document TMail extra components (#272)
linagora/tmail-backend: docs/modules/ROOT/pages/tmail-backend/extra-components/index.adoc

= TMail optional components
:navtitle: Optional components

TMail offers some alternatives to James components, solving some specific pain points.

== Object Storage: BlobId list

> When deduplication upon blob storage is enabled in James, we would store only one objects for all

Reviewer: ```suggestion
When deduplication upon blob storage is enabled in James, we would store only one object for all
```

entities sharing the same content.

This helps reducing storage costs as email content within companies is highly duplicated.

However, this brings additional issues and opportunity:

>  - We might not want to save several time the same content, when we know we actually already store it. Storing it twice

Reviewer: ```suggestion
 - We might not want to save several times the same content, when we know we actually already store it. Storing it twice
```

is a waste of resources and writes might be billed by the cloud provider.
 - Concurrently reading and writing to the same blob might cause issues for low consistency object stores (eg Swift based),
see link:https://issues.apache.org/jira/browse/JAMES-3570[this issue].

We thus offer an additional cache, in Cassandra, to check if the blob is already stored prior to store it. Inconsistent

> cache result in double writes (low impact). We found it to have a positive impact on performance and solve isolation issues

Reviewer: ```suggestion
cache results in double writes (low impact). We found it to have a positive impact on performance and solve isolation issues
```

on top of Swift S3 APIs.

Example `blob.properties`:

....
[...]
# When deduplication is active, use a Cassandra projection to avoid saving several time the same blob to S3.
# This prevents isolation issues for Swift backed S3 deployments, also adds a performance boost.
# Mandatory, Allowed values are: true, false
single.save.enable=false
[...]
....

== Combined user repository

Linagora uses LDAP as an identity provider.

We use it in a couple of occasions including:

>  - IMAP / SMTP login, which involve checking the user password against the LDAP

Reviewer: ```suggestion
 - IMAP / SMTP login, which involves checking the user password against the LDAP
```

>  - Receiving an email (both at the SMTP and mail processing level, we check that the user exist)

Reviewer: ```suggestion
 - Receiving an email (both at the SMTP and mail processing level, we check that the user exists)
```

With such a set up, we need to plan a recurring synchronisation task using link:https://github.com/lsc-project/lsc-james-plugin#users-synchronization[LSC].
See link:https://github.com/lsc-project/lsc-james-plugin/tree/master/sample/ldap-to-james-user[this sample configuration]

Also, we await a higher availability as emails can still be received during LDAP outage. The only impact will only be no new SMTP/IMAP login.

Furthermore, the load on a LDAP will be reduced to LSC synchronisation and IMAP/SMTP logins.

Example `usersrepository.xml`:

....
<usersrepository name="LocalUsers"
    class="org.apache.james.user.ldap.CombinedUsersRepository"
    ldapHost="${env:OP_JAMES_LDAP_URI}"
    userBase="${env:OP_JAMES_LDAP_BASE_DN}"
    principal="${env:OP_JAMES_LDAP_PRINCIPAL_DN}"
    credentials="${env:OP_JAMES_LDAP_BIND_PW}"
    userIdAttribute="mail"
    userObjectClass="person"
    administratorId="${env:OP_JAMES_ADMINISTRATOR_ID}" >
    <enableVirtualHosting>true</enableVirtualHosting>
</usersrepository>

....

== Enabling/Disabling the encrypted mailbox

Encrypting messages is costly thus we unserstand an administrator might choose not to enable this feature.

> We provide an easy way to disactivate completly the **Encrypted Mailbox** through the configuration.

Reviewer: ```suggestion
We provide an easy way to disable completely the **Encrypted Mailbox** through the configuration.
```
